### ISSUE TITLE
support redirection on source IP

### ISSUE THREAD

**Issue by xaleeks** [projectcontour/contour#3789]
There’s an interesting requirement from k8s-infra-WG to use an Envoy based solution for doing 302 redirects to different instances of an application running on various public clouds (as well as baremetal).  This ticket is really for us to explore this idea and see if Contour is an appropriate solution.  They are using BGP ASNs to determine the correct instance to route to which is based on a variety of factors but mostly on proximity, ie. client request from Germany goes to Berlin, USwest requests goes to San Fran, APAC goes to Singapore etc.  They’ve experimented with basic Envoy filter for writing redirects and looking for some additional validation.  If we have Contour being that re-director, basic flow might be 

1. Client always reaches out to the same url
2. This request is received by a dedicated Contour ingress controller running somewhere
3. Redirects request to the correct instance based on source IP (leveraging our support for ExternalName to route externally)
4. Requests can be allowed / blocked based on source IP info 
5. 

So Contour needs to be able to distinguish requests based on source IP coming from different ASNs ( amazon, google, aws, digital ocean, etc) 

can see this touching on a couple of things in our backlog

- LUA filter support
- Source IP affinity https://github.com/projectcontour/contour/issues/3703     

Q:
What’s advantage of using this vs some dedicated web server for doing 302s, of which there are plenty of solutions on the market?  

thoughts? @youngnick  @stevesloka

**Comment by youngnick** [comment]
I think that this use case is actually very tricky to implement in Contour. The problems I see are:
- what makes the routing decision, based on the source IP? If it's Contour, we need a way to manage the subnets that will be used for routing. This does not seem like a core competency for Contour, and feels like something that would be better delegated to an external Envoy filter of some sort (similar to Rate Limiting).
- How do you manage the source IP ranges? Filtering by ASN means that you will be storing many, many thousands or even millions of subnets, with associated ASN paths. This will require either participating in BGP or looking these routes up using a looking glass of some sort. In either case, this is a sufficiently complex problem to be worthy of being a project in its own right, and is probably not a core competency for the Contour team as it stands.

In short, I think that Contour could be a *component* in this sort of solution, but I think that this is a large enough problem it's outside the scope of Contour (the controller).
